### PR TITLE
Week5 Analytics Homework Solution

### PR DESCRIPTION
**Pull request by AliaHa3**

### CODE REVIEW

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

The student's solution is well-structured and logically follows the given homework prompt. The student has correctly implemented the state change tracking for `nba_players` as per the requirements. The use of Common Table Expressions (CTEs) to break down the problem into smaller, more manageable parts is commendable. 

However, there are a few areas where the solution could be improved:

1) **Readability and Comments:** The student's code is generally readable, but it could be improved by adding more comments to explain the logic behind each step, especially for complex queries. For example, the logic behind the `CASE` statement in the final `SELECT` query could be explained in more detail.

2) **Error Handling:** The student's code does not seem to handle potential errors or edge cases. For example, what happens if the `player_name` is `NULL` or if the `current_season` is not a valid year? Adding error handling or input validation could make the code more robust.

3) **Modularity and Reusability:** The student's code is modular and each CTE can be understood independently. However, the code could be made more reusable by parameterizing the year, instead of hardcoding it as `1996`. This would allow the same code to be used for any year.

4) **Code Efficiency:** The student's code could be made more efficient by avoiding unnecessary computations. For example, the `active_season - first_active_season` and `active_season - last_active_season` calculations are repeated multiple times in the final `SELECT` query. These could be calculated once in the `combined` CTE and reused in the final `SELECT` query.

Here's a suggestion for the final `SELECT` query:

```sql
SELECT
  player_name,
  first_active_season,
  last_active_season,
  seasons_active,
  CASE
    WHEN active_season_diff_first = 0 THEN 'New'
    WHEN active_season_diff_last > 1 THEN 'Returned from Retirement'
    WHEN active_season_diff_last = 0 THEN 'Continued Playing'
    WHEN active_season IS NULL
    AND partition_season_diff_last = 1 THEN 'Retired'
    ELSE 'Stayed Retired'
  END AS season_active_state,
  partition_season
FROM
  (
    SELECT *,
      active_season - first_active_season AS active_season_diff_first,
      active_season - last_active_last_season AS active_season_diff_last,
      partition_season - last_active_season AS partition_season_diff_last
    FROM combined
  )
```

Overall, the student's solution is on the right track. With a few improvements, it could be made more readable, robust, and efficient.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used the `GROUPING SETS` function to aggregate the data along the required dimensions. However, there are a few areas where your solution could be improved.

1) **Missing Aggregation**: The homework prompt asks for the sum of `pts` and `won` for each grouping set. However, your solution only includes the sum of `pts`. You need to add the sum of `won` to your query. You can do this by adding `SUM(won) AS total_team_wins` to your `SELECT` clause, similar to the example solution.

2) **Data Deduplication**: The example solution includes a `WITH` clause at the beginning to deduplicate the data. This is an important step to ensure accurate results. Your solution seems to be using a pre-deduplicated table `nba_game_details_dedup`. If this table is not properly deduplicated, it could lead to incorrect results. Make sure you understand how this table was created and that it is indeed deduplicated.

3) **Error Handling**: Your solution includes a `CASE` statement to handle null values in the `pts` column. This is a good practice for robustness and error handling. However, you should also consider handling potential null values in other columns that are used in your query.

4) **Readability**: Your code is generally well-formatted and easy to read. However, you could improve readability by adding comments to explain your thought process and the purpose of each part of your query. This will make it easier for others (and future you) to understand your code.

5) **Modularity and Reusability**: Your code is structured in a way that makes it easy to understand and reuse specific components in other contexts. However, you could improve the modularity of your code by creating a separate function or view for the deduplication process, especially if this is a step that needs to be repeated often.

In summary, your solution is on the right track, but there are a few areas where it could be improved. Keep up the good work!

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used `GROUPING SETS` to perform aggregations of the `game_details` data along the required dimensions. Your code is also quite readable, with meaningful variable names and clear comments explaining your thought process. 

However, there are a few areas where your solution could be improved:

1) **Missing `won` column:** The homework prompt asks for the aggregation of both points (`pts`) and wins (`won`). Your solution currently only aggregates points. You should add a column to your `SELECT` statement that sums the `won` column, similar to the example solution.

2) **Error handling:** Your solution uses `COALESCE` to handle null values in the `player_name`, `team_abbreviation`, and `season` columns, which is good. However, you should also consider handling potential errors that could arise from the `JOIN` operation. For example, you could add a `WHERE` clause to filter out rows where `game_id` is null before performing the join.

3) **Modularity and reusability:** Your solution is quite modular and reusable, but it could be improved by defining the `JOIN` operation as a separate Common Table Expression (CTE), similar to the `deduped` CTE in the example solution. This would make your code more modular and easier to understand.

4) **Formatting for readability:** While your code is generally well-formatted, it could be improved by adding more whitespace and indentation to make it easier to read. For example, you could indent the `GROUPING SETS` clause and the `ORDER BY` clause to align with the `FROM` clause.

Here's how you could revise your solution to address these points:

```sql
-- This query calculates total points scored and total wins in NBA games, providing subtotals and overall totals.
WITH joined_data AS (
    SELECT 
        ngd.*,
        ng.season,
        CASE 
            WHEN ng.visitor_team_id = ngd.team_id AND ng.home_team_wins = 1 THEN 0
            WHEN ng.visitor_team_id = ngd.team_id AND ng.home_team_wins = 0 THEN 1
            ELSE ng.home_team_wins
        END AS won
    FROM 
        bootcamp.nba_game_details_dedup ngd -- using pre-deduped table that exist in bootcamp schema
        INNER JOIN bootcamp.nba_games ng ON ngd.game_id = ng.game_id
    WHERE 
        ngd.game_id IS NOT NULL
)
SELECT
    -- Use COALESCE to handle nulls, defaulting to 'Overall' for aggregate rows
    COALESCE(joined_data.player_name, 'Overall') as player,
    COALESCE(joined_data.team_abbreviation, 'Overall') as team,
    COALESCE(CAST(joined_data.season AS VARCHAR), 'Overall') as season,
    -- Sum points, treating nulls as 0
    SUM(
        CASE
            WHEN joined_data.pts IS NULL THEN 0
            ELSE joined_data.pts
        END
    ) AS total_points,
    -- Sum wins, treating nulls as 0
    SUM(
        CASE
            WHEN joined_data.won IS NULL THEN 0
            ELSE joined_data.won
        END
    ) AS total_wins
FROM
    joined_data
GROUP BY
    -- Grouping sets to create subtotals for each player, team, and season
    GROUPING SETS (
        (joined_data.player_name, joined_data.team_abbreviation),
        (joined_data.player_name, joined_data.season),
        (joined_data.team_abbreviation)
    )
ORDER BY
    -- Order results by player, team, and season
    player,
    team,
    season
```

Keep up the good work! With a few minor adjustments, your solution will be even better.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your code is well-structured, readable, and logically follows the homework prompt. You've done a great job using `GROUPING SETS` to perform aggregations of the `game_details` data. However, there are a few areas where your solution could be improved.

1) **Covering all aspects of the homework prompt:** Your code does not calculate the total team wins, which is a requirement in the prompt. You can add this by including a `SUM(won) AS total_team_wins` in your `SELECT` statement, similar to the example solution.

2) **Readability:** Your code is generally readable, with meaningful variable names and clear comments explaining your thought process. However, you could improve the readability by adding more comments to explain complex parts of your code, such as the `GROUPING SETS` clause.

3) **Robustness and error handling:** Your code handles null values in the `pts` column by treating them as 0, which is a good practice. However, it does not handle potential errors in the `JOIN` clause. You could improve this by adding error handling code to check if the `JOIN` operation is successful.

4) **Modularity and reusability:** Your code is modular and reusable, with a clear separation of concerns between the `WITH` clause and the main query. However, you could improve the reusability of your code by defining the `WITH` clause as a separate function that can be reused in other queries.

5) **Overall structure and coherence:** Your code is well-structured and logically follows the homework prompt. However, you could improve the coherence of your code by aligning the `SELECT` and `FROM` clauses with the `WITH` clause, similar to the example solution.

Keep up the good work, and with a few improvements, your code will be even better!

### Feedback for `query_3.sql`

Overall, your solution is well-structured and demonstrates a good understanding of SQL and the use of `GROUPING SETS`. However, there are a few areas where it could be improved to better align with the assignment prompt and to enhance readability and efficiency.

1) **Alignment with the prompt:** The assignment prompt asks for a query that answers the question: "Which player scored the most points playing for a single team?" Your solution seems to be doing more than that. It calculates the total points scored in NBA games, providing subtotals and overall totals. While this is a more comprehensive approach, it may not be necessary for the specific question asked in the prompt. The example solution is more straightforward and directly answers the question.

2) **Readability:** Your code is generally well-structured and includes comments, which is good practice. However, the complexity of your solution makes it harder to understand. The use of multiple CTEs (Common Table Expressions) and the `GROUPING SETS` function may not be necessary to answer the question. A simpler solution, like the example provided, would be easier to read and understand.

3) **Efficiency:** Your solution involves multiple joins and aggregations, which could be computationally expensive. The example solution is more efficient as it directly queries the required data without additional processing.

4) **Error handling and robustness:** Your code handles null values well using the `COALESCE` function and conditional logic in the `SUM` function. However, it's not clear if your code handles other potential errors or edge cases, such as negative point values or missing player or team names.

5) **Modularity and reusability:** While your code is modular and could potentially be reused in other contexts, it may be overly complex for this specific task. The example solution is simpler and more focused, making it easier to adapt for similar tasks.

Here's a suggestion for improvement based on the example solution:

```sql
SELECT	player_name, team_abbreviation, SUM(pts) as total_player_points
FROM game_details_cte
WHERE pts IS NOT NULL 
GROUP BY player_name, team_abbreviation
ORDER BY total_player_points DESC
LIMIT 1
```

This query directly calculates the total points for each player-team combination and returns the one with the highest total points. It's simpler, more efficient, and directly answers the question in the prompt.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

Overall, your solution is well-structured and logically follows the homework prompt. You've successfully written a query to answer the question: "Which player scored the most points playing for a single team?" However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution is more complex than necessary for the task at hand. The prompt asks for a simple query on top of the results of the `GROUPING SETS` aggregations, but your solution includes additional CTEs and aggregations that are not required. The example solution demonstrates a more straightforward approach to this task.

2) **Readability:** Your code is generally well-formatted and includes meaningful variable names, which makes it easier to understand. However, the complexity of your solution makes it harder to follow than it needs to be. Simplifying your code would improve its readability.

3) **Robustness and error handling:** Your solution includes good error handling, such as the use of `COALESCE` to handle null values. However, it's not clear why you're treating null points as 0 in the `nba_game_details_dashboard` CTE. If a player's points are null, it might be more accurate to exclude that row from the sum rather than treating it as 0.

4) **Modularity and reusability:** Your use of CTEs makes your code modular and potentially reusable. However, the complexity of your solution might make it harder to adapt for other tasks. Simplifying your code would make it more reusable.

Here's a suggestion for a simpler solution based on the example:

```sql
SELECT	player_name, team_abbreviation, SUM(pts) as total_player_points
FROM game_details_cte
WHERE pts IS NOT NULL 
GROUP BY player_name, team_abbreviation
ORDER BY total_player_points DESC
LIMIT 1
```

This query sums the points for each player-team combination and returns the combination with the highest total points. It's simpler than your solution but achieves the same result.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.
